File: 0001-0500/0322.Coin Change.java
import java.util.Arrays;
public class Solution {
    public int coinChange(int[] coins, int amount) {
        int dp[] = new int[amount+1];
        Arrays.fill(dp,Integer.MAX_VALUE);
        dp[0]=0;
        for(int i=1;i<=amount;i++)
            for(int coin : coins)
                if(i-coin>=0 && dp[i-coin]!=Integer.MAX_VALUE && dp[i-coin]+1<dp[i])
                    dp[i]=dp[i-coin]+1;
        if(dp[amount]==Integer.MAX_VALUE)
            return -1;
        return dp[amount];
    }
}

File: 0001-0500/0322.Coin Change.py
"""
Basic idea:
	Dynamic Programming
	Store the least number of coins required to make up i, for each i<amount

180 / 180 test cases passed.
Status: Accepted
Runtime: 912 ms
Sorry. We do not have enough accepted submissions.
"""

class Solution(object):
    def coinChange(self, coins, amount):
        """
        :type coins: List[int]
        :type amount: int
        :rtype: int
        """
        coins.sort()    #sort for pruning when i-c<0
        table = [amount+1]*(amount+1)    #table[i-1] indicates the least number of coins to makeup amount i
        table[0]=0
        for i in xrange(1,amount+1):
            for c in coins:
                if i-c<0:
                    break
                if table[i-c]+1<table[i]:
                    table[i]=table[i-c]+1
        return table[amount] if table[amount]<=amount else -1

